File: 2022/08.py
from utils.grid import NEIGH_DIRECT


def get_input(puzzle: str) -> list[list[int]]:
    return [[*map(int, line)] for line in puzzle.splitlines()]


def part1(puzzle: str):
    grid = get_input(puzzle)
    w, h = len(grid[0]), len(grid)
    visible = set()
    for dx, dy in NEIGH_DIRECT:
        for y in range(h):
            for x in range(w):
                i = y + dy
                j = x + dx
                while i in range(h) and j in range(w):
                    if grid[i][j] >= grid[y][x]:
                        break
                    i += dy
                    j += dx
                else:
                    visible.add((x, y))
    return len(visible)


def part2(puzzle: str):
    grid = get_input(puzzle)
    out = 0
    for y in range(len(grid)):
        for x in range(len(grid[y])):
            d = 1
            for dx, dy in NEIGH_DIRECT:
                i = y + dy
                j = x + dx
                s = 0
                while i in range(len(grid)) and j in range(len(grid[0])):
                    s += 1
                    if grid[i][j] >= grid[y][x]:
                        break
                    i += dy
                    j += dx
                d *= s
            out = max(out, d)
    return out


if __name__ == "__main__":
    from aoc import run

    run(2022, 8, part1, part2)

File: 2022/08.rs
use aoc::{grid::Direction, iter_ext::IterExt};

type Input = Vec<Vec<u8>>;

struct CoordIterator<'a> {
    x: usize,
    y: usize,
    width: usize,
    height: usize,
    direction: &'a Direction,
}

impl Iterator for CoordIterator<'_> {
    type Item = (usize, usize);

    fn next(&mut self) -> Option<Self::Item> {
        (self.x, self.y) = self
            .direction
            .step(self.x, self.y, self.width, self.height)?;
        Some((self.x, self.y))
    }
}

fn setup(input: &str) -> Input {
    input
        .trim()
        .lines()
        .map(|line| line.bytes().map(|x| x - b'0').collect())
        .collect()
}

fn part1(grid: &Input) -> usize {
    let (height, width) = (grid.len(), grid[0].len());

    (0..height)
        .flat_map(|y| {
            (0..width).filter(move |&x| {
                Direction::iter().any(|direction| {
                    CoordIterator {
                        x,
                        y,
                        width,
                        height,
                        direction,
                    }
                    .all(|(j, i)| grid[y][x] > grid[i][j])
                })
            })
        })
        .count()
}

fn part2(grid: &Input) -> usize {
    let (height, width) = (grid.len(), grid[0].len());

    (0..height)
        .flat_map(|y| {
            (0..width).map(move |x| {
                Direction::iter()
                    .map(|direction| {
                        CoordIterator {
                            x,
                            y,
                            width,
                            height,
                            direction,
                        }
                        .take_while_inclusive(|&(j, i)| grid[y][x] > grid[i][j])
                        .count()
                    })
                    .product::<usize>()
            })
        })
        .max()
        .unwrap()
}

aoc::main!(2022, 8);
aoc::example!(ex01, "08.1.txt", 21, 8);
aoc::test_input!("08.txt", 1825, 235200);
